### PR TITLE
Bound FastMCP injection claims to whole-signature evidence

### LINKED ISSUE #542: Bound FastMCP injection claims to the SDK's whole-signature and generic-type rules
## Delivered — 2026-09-09

Completed in #603, merged `5c5a4b94118c93e2493812689d3920212451f1e4`. The installed/shared idiom reader and zero-install port now make a bounded whole-signature injection decision, remove at most the first proven Context, and retain named signature/parameter limits where generic, annotation or class evidence is unresolved. Canonical/nullable Context, local proven subclasses and ordinary application models retain their supported behavior. No target server is imported or executed; medium confidence and tool discovery are preserved.

The published support note pins SDK 1.27.2 and standalone FastMCP 2.14.5 source behavior and explicitly does not claim full runtime fidelity. Framework-family/exact-export provenance is separately filed and deferred in #601.

Two independent GitHub coding-agent review/address rounds (reviews 5149304725 and 5149387533; addresses 5595251733 and 5595278042) completed. Round 1 found invalid typing arity/empty Union could hide Context; commit `9493d77545e8756e11daebd17d67d7dc67490967` fixed it with SDK-backed red/green probes and shared production/standalone coverage. Final full suite **9,203 passed, 5 skipped**; round-2 reviewer independently passed 690 checks. All final CI including aggregate coverage and fresh current-base verifier authorization passed. An initial automated merge rejection cited pending checks; direct re-read evidence showed them complete and the same merge interface then succeeded. This does not establish v1.0 qualification.

## v1.0 release disposition — 2026-09-08

v1.0 sequencing (#572): settle and verify the SDK profiles actually promised or consumed by qualification before contract freeze. A bounded support repair or explicit unsupported limit is the release slice; a general Python type evaluator is not required.

Found during review of #541 / #539; deferred from that implementation at the user's request.

## Evidence

The supported MCP SDK's `find_context_parameter` inspects the whole function, returns the first matching parameter, and examines immediate generic arguments for a Context subclass. With the installed SDK, a synthetic `def generic(holder: list[Context]) -> str` returns `holder` as the injected parameter. The existing loader test instead says `list[Context]` is a caller input. The reader classifies each parameter independently and can therefore also disagree with first-match behavior or with the SDK's all-or-nothing `get_type_hints` failure.

This is a framework-semantic mismatch, not a claim that an unsafe release was reproduced. The two Context identity bugs and nullable evidence repair remain in #539.

## Proposed follow-up

1. Pin the supported SDK/framework-family injection semantics using their primary source and small synthetic signature fixtures; standalone FastMCP and the SDK must not be assumed identical.
2. Cover generic Context arguments, two Context parameters, an unresolved annotation elsewhere in the signature, imported subclasses/re-exports, and normal application models.
3. Publish a named limitation where static evidence cannot establish the whole-signature result; retain tool discovery and the medium ceiling. Do not execute a target server or import its modules.
4. Reuse the shared idiom corpus and installed/standalone parity tests. Do not build a general Python type evaluator merely to close this edge case.

## Acceptance

- [x] Generic and first-match fixtures agree with the specified framework semantics, or explicitly retain a local coverage limit.
- [x] Unresolved external class provenance is not described as established caller ownership solely because its import path lies outside framework packages.
- [x] Canonical Context, ordinary context: str and application Context(BaseModel) preserve their established behavior.
- [x] The documented support boundary names any remaining disagreement, without claiming full injection fidelity.

Primary source: https://github.com/modelcontextprotocol/python-sdk/blob/v1.x/src/mcp/server/fastmcp/utilities/context_injection.py . Main implementation audited at PR #541 commit 89f4967736b8db6b97c1435cf90b6bb1737517e9.

## Priority and scope

P2 bounded correctness follow-up. Deferred because resolving all SDK/framework-family injection semantics expands the current #539 repair. No new verdict, adapter or runtime path.

### PR DESCRIPTION
**Pull request by pengfei-threemoonslab**
FastMCP injection was inferred per parameter, so multiple Context parameters could disappear from a tool schema even though both pinned framework helpers select only the first match. Unresolved annotations elsewhere in the signature and generic Context shapes were also treated as stronger evidence than the local reader could establish.

The shared static reader now evaluates the whole signature before claiming injection, excludes at most the first proven Context, and keeps uncertain parameters visible with an explicit function-level coverage limit. The documented common profile retains direct/nullable Context and proven local subclasses; parameterized or nested Context, unknown annotations/metadata, unresolved class construction and external class exports remain bounded limits. The standalone detector carries the same implementation and parity assertions. No scanned server is imported or executed.

Closes #542 under #572. Framework-family and exact-export provenance is separately recorded and deferred in #601; this PR does not claim full SDK or standalone FastMCP runtime fidelity.

## Validation

- Actual installed SDK 1.27.2 utility probes, shared first-match/whole-signature fixtures, local model and unresolved-class cases.
- 1,046 focused source/standalone/distribution tests passed, 1 skipped; real bare-stdlib and CRLF detector coverage included.
- Initial final suite: 9,186 passed, 5 skipped. After the round-1 arity correction: 9,203 passed, 5 skipped. The initial full run caught stale standalone parity; the port was corrected before this final run.
- Whole-tree Ruff and diff checks passed. Committed verification against current main passed with fresh control `complete`.

## Review

Round-1 COMMENT review 5149304725 found a known-typing-arity/empty-Union gap. Commit `9493d77545e8756e11daebd17d67d7dc67490967` addresses it with SDK-backed red/green regressions and shared loader/standalone coverage. Round 2 and final-head CI remain required before merge. This is an adapter support correction and does not establish v1.0 qualification.

### CODE REVIEW

File: src/agents_shipgate/inputs/mcp_idioms.py [hunk 1]
```diff
@@ -2889,32 +2932,66 @@ def _python_context_injection(
...
+            if len(elements) < 2 or not all(isinstance(item, ast.Constant) for item in elements[1:]):
+                return "unresolved"
+            return resolve(elements[0])
+        if symbol == "Literal":
+            return "caller_supplied" if all(isinstance(item, ast.Constant) for item in elements) else "unresolved"
+        if symbol in {"Union", "Optional"}:
+            if symbol == "Optional" and len(elements) != 1:
+                return "unresolved"
+            identities = {resolve(item) for item in elements}
+        elif symbol in {"list", "List", "dict", "Dict", "tuple", "Tuple", "set", "Set", "frozenset", "FrozenSet"}:
```
Comment (pengfei-threemoonslab) [inline src/agents_shipgate/inputs/mcp_idioms.py:2982]: [P2] Reject malformed known typing arities before claiming whole-signature injection. For a deferred signature ctx: Context, payload: List[str, int], installed SDK 1.27.2 find_context_parameter returns None because get_type_hints fails; this branch treats both generic arguments as caller-supplied, so the production loader drops ctx and reports an enumerated surface with no gap. Dict[str] and empty Union[()] similarly drop ctx without unresolved_context_signature. Check the fixed arities of the known typing aliases and reject an empty Union (or retain an explicit local limit) before reducing member identities; keep builtin generic behavior separate. Add installed-SDK refusal probes and shared/standalone/production regressions asserting ctx remains visible with the whole-signature gap. This is the existing #542 acceptance boundary, not the framework-export provenance work deferred in #601.

**Review by pengfei-threemoonslab — COMMENTED** [review]
Independent coding-agent review, round 1, against head 87692c2886670e5eedf534798d44f906ae355d3b and base e6f754512cbc5c09d9472cf294670e487de1ed16. This is a coding-agent COMMENT, not human approval or merge/release authority.

One actionable P2 finding: invalid arity in a known typing alias can still conceal the Context parameter. See the inline comment on mcp_idioms.py:2982. Installed MCP SDK 1.27.2 refuses whole-signature resolution for deferred List[str, int], Dict[str] and Union[()] annotations, returning no injected parameter; the reader instead classifies the annotation as caller-owned and excludes ctx. In the actual production loader, List[str, int] produces parameters=[payload], surface=enumerated and no surface gaps. Preserve an explicit whole-signature limit for these malformed known typing forms; a general Python type evaluator is unnecessary.

Verified all nine published GitHub blob SHAs against the clean frozen checkout. Independently passed the 336 tests in the new injection contract, shared idiom and production-loader families, plus 337 shared-reader/standalone corpus and CRLF parity cases. Additional synthetic utility probes against installed SDK 1.27.2 agree for first-match selection, unknown return/variadic refusal, local Context subclasses and ordinary Pydantic application models. Only self-authored fixture signatures were inspected by the installed SDK utility; no scanned server was imported or executed.

The function-level coverage gap, retained unknown parameters, medium confidence ceiling and standalone port are otherwise consistent with the stated bounded common profile. The remaining framework-family/exact-export provenance problem is explicitly documented and remains deferred in #601; this review does not promote it into this PR or claim full framework/runtime fidelity or v1.0 qualification.

**Comment by pengfei-threemoonslab** [comment]
Author address — round 1, now at `9493d77545e8756e11daebd17d67d7dc67490967`.

Fixed P2 from [review 5149304725](https://github.com/ThreeMoonsLab/agents-shipgate/pull/603#pullrequestreview-5149304725): known typing aliases now enforce their construction arity (`List`/`Set`/`FrozenSet`: one, `Dict`: two), and an empty Union is unresolved before any Context exclusion. Builtin GenericAlias behavior remains distinct. Both installed and standalone readers received the same narrow correction.

Six newly authored SDK-backed regressions failed on the reviewed head before the fix. The SDK probes establish actual `get_type_hints` TypeError/no-injection behavior; shared corpus and production-loader tests assert that Context remains visible with a signature coverage gap. Valid-arity and existing builtin-container controls retain their behavior. All targeted families passed; full final suite **9,203 passed, 5 skipped**. Ruff/diff checks and current-main committed verification passed, with fresh current control `complete`.

Round 2 independent review is requested on this exact commit. #601 remains separately deferred; no new runtime-fidelity claim, policy override or qualification claim is made.

**Review by pengfei-threemoonslab — COMMENTED** [review]
Independent coding-agent review, round 2, against head 9493d77545e8756e11daebd17d67d7dc67490967 and base e6f754512cbc5c09d9472cf294670e487de1ed16. This is a coding-agent COMMENT, not human approval or merge/release authority.

The round-1 P2 finding is addressed. Both readers reject incorrect fixed arities for the recognized typing List/Set/FrozenSet/Dict aliases and an empty Union before reducing member identities. The whole-signature limit now retains ctx for the reproduced SDK refusal cases; builtin GenericAlias behavior remains separate. I found no remaining actionable defect in the reviewed change.

Verified all nine published GitHub blob SHAs against the clean frozen checkout and inspected the six-file address delta. Independently passed 351 injection-contract/shared-reader/production-loader tests and 339 shared-reader/standalone corpus and CRLF parity cases (690 total). Additional probes against installed MCP SDK 1.27.2 confirmed all six valid controls, including builtin list[str, int]; qualified and aliased malformed List annotations now remain unresolved, as do malformed return and variadic annotations. The production-loader regression asserts both parameters remain visible with the function-level coverage limitation.

No scanned server was imported or executed, and no implementation source or control artifacts were modified. The bounded common profile, medium confidence ceiling and explicit unknown evidence remain intact. Framework-family and exact-export provenance stays deferred in #601; this does not claim full framework fidelity, runtime safety or v1.0 qualification. No third review round is requested for this unchanged head.

**Comment by pengfei-threemoonslab** [comment]
Author address — round 2, unchanged head `9493d77545e8756e11daebd17d67d7dc67490967`.

Independent COMMENT [5149387533](https://github.com/ThreeMoonsLab/agents-shipgate/pull/603#pullrequestreview-5149387533) confirms the round-1 P2 is fixed and reports no remaining actionable in-scope findings. The reviewer independently passed 690 reader/loader/contract and standalone/CRLF checks, matched all nine published blobs, and probed valid/builtin controls plus malformed qualified/aliased typing forms, returns and variadics against SDK 1.27.2. No further source change is needed.

Two coding-agent review/address rounds are complete. Final-head CI/aggregate coverage and a fresh current-base control read remain required for merge. #601 remains deferred; neither these reviews nor this support correction establish independent human approval or v1.0 qualification.
